spin/cmds/meson.py

```diff
@@ -351,30 +362,37 @@ def build(
 
         # Any other conditions that warrant a reconfigure?
 
+    meson_args_compile = list(meson_args.get("compile", tuple()))
     compile_flags = ["-v"] if verbose else []
-    if jobs:
+    if "jobs" in meson_args:
+        jobs = meson_args["jobs"]
+    if jobs is not None:
         compile_flags += ["-j", str(jobs)]
 
     p = _run(
-        _meson_cli() + ["compile"] + compile_flags + ["-C", build_dir],
+        _meson_cli() + ["compile"] + compile_flags + ["-C", abs_build_dir] + meson_args_compile,
         sys_exit=True,
         output=not quiet,
     )
-    p = _run(
-        _meson_cli()
-        + [
+
+    meson_args_install = list(meson_args.get("install", tuple()))
+    cmd = _meson_cli() + [
             "install",
             "--only-changed",
             "-C",
-            build_dir,
-            "--destdir",
-            install_dir
-            if os.path.isabs(install_dir)
-            else os.path.relpath(abs_install_dir, abs_build_dir),
-        ],
-        output=(not quiet) and verbose,
-    )
+            build_dir
+        ] + meson_args_install
```

Comment:
This also may or may not be backwards compatible. The change made here is removal of `--destdir`. I think with this the `--prefix` will be used as installation directory.

Let me know if I am mis-interpreting things.

Reply:
I think you don't need these changes. What goes wrong if you don't touch this code at all? If the only thing that changes is the install path going from this:
```
build-install/lib/python3.12/site-packages/scipy/
```
to:
```
build-install/usr/lib/python3.12/site-packages/scipy/
```
then that is fine, that's a completely harmless change.

Reply:
Actually I we follow Option 1 from https://github.com/scientific-python/spin/pull/238#discussion_r1834424443 and undo this change then from my last experience, `spin test` didn't work correctly. Actually I made the changes here to match with the series of commands executed by `python dev.py build`. Then everything works in SciPy (also with NumPy). I can try undoing this changes and let you know if SciPy still works.

Reply:
Here is the error if I undo my changes,

```zsh
(scipy-dev) 20:11:53:~/Quansight/scipy % spin test -s stats
Invoking `build` prior to running tests:
$ meson compile -j 10 -C /Users/czgdp1807/Quansight/scipy/build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /Users/czgdp1807/mambaforge/envs/scipy-dev/bin/ninja -C /Users/czgdp1807/Quansight/scipy/build -j 10
ninja: Entering directory `/Users/czgdp1807/Quansight/scipy/build'
[1/1] Generating scipy/generate-version with a custom command
$ meson install --only-changed -C build --destdir ../build-install --tags=runtime,python-runtime,tests,devel
$ export PYTHONPATH="/Users/czgdp1807/Quansight/scipy/build-install/usr/lib/python3.12/site-packages"
$ /Users/czgdp1807/mambaforge/envs/scipy-dev/bin/python3.12 -P -c 'import scipy'
$ cd /Users/czgdp1807/Quansight/scipy/build-install/usr/lib/python3.12/site-packages
$ /Users/czgdp1807/mambaforge/envs/scipy-dev/bin/python3.12 -P -m pytest --import-mode=importlib --pyargs scipy.stats -m 'not slow'
===================================================== test session starts ======================================================
platform darwin -- Python 3.12.6, pytest-8.3.3, pluggy-1.5.0
rootdir: /Users/czgdp1807/Quansight/scipy
configfile: pytest.ini
plugins: cov-5.0.0, anyio-4.4.0, hypothesis-6.112.1, timeout-2.3.1, xdist-3.6.1
collected 0 items / 35 errors                                                                                                  

============================================================ ERRORS ============================================================
___________________ ERROR collecting build-install/usr/lib/python3.12/site-packages/scipy/stats/_bws_test.py ___________________
ImportError while importing test module '/Users/czgdp1807/Quansight/scipy/build-install/usr/lib/python3.12/site-packages/scipy/stats/_bws_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
scipy/stats/__init__.py:614: in <module>
    from ._stats_py import *
scipy/stats/_stats_py.py:39: in <module>
    from scipy.spatial import distance_matrix
scipy/spatial/__init__.py:116: in <module>
    from ._geometric_slerp import geometric_slerp
scipy/spatial/_geometric_slerp.py:9: in <module>
    from scipy.spatial.distance import euclidean
scipy/spatial/distance.py:121: in <module>
    from ..special import rel_entr
scipy/special/__init__.py:816: in <module>
    from . import _ufuncs
E   ImportError: dlopen(/Users/czgdp1807/Quansight/scipy/build-install/usr/lib/python3.12/site-packages/scipy/special/_ufuncs.cpython-312-darwin.so, 0x0002): Library not loaded: /usr/lib/python3.12/site-packages/scipy/special/libsf_error_state.dylib
E     Referenced from: <C40E1CEC-A9D0-39F7-8466-1ABD39B3EF5D> /Users/czgdp1807/Quansight/scipy/build-install/usr/lib/python3.12/site-packages/scipy/special/_ufuncs.cpython-312-darwin.so
E     Reason: tried: '/usr/lib/python3.12/site-packages/scipy/special/libsf_error_state.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/python3.12/site-packages/scipy/special/libsf_error_state.dylib' (no such file), '/usr/lib/python3.12/site-packages/scipy/special/libsf_error_state.dylib' (no such file, not in dyld cache)
```

Reply:
The shared library bites again:) No worries, I'm pretty sure I can fix that. I'll try to have a look later today.

To help me here, could you point me to a different `spin` branch you have that allows me to reproduce this problem from a `spin test` in a clean repo?

Reply:
Yes. `spin_upd_1` (the branch which contains the changes I reverted from here). Link - https://github.com/czgdp1807/spin/tree/spin_upd_1

Reply:
It's the same issue as is being discussed in https://github.com/numpy/numpy/pull/27371, it has nothing to do with the shared library but with a breaking change in `spin` for test discovery that happened in a recent release.

E.g., for `spin test -s stats`:
```
ImportError while importing test module '/home/rgommers/code/scipy/build-install/usr/lib/python3.12/site-packages/scipy/stats/_page_trend_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
scipy/stats/_page_trend_test.py:4: in <module>
    from ._continuous_distns import norm
E   ModuleNotFoundError: No module named 'build-install.usr.lib.python3.12.site-packages.scipy.stats._continuous_distns'; 'build-install.usr.lib.python3.12.site-packages.scipy.stats' is not a package
...
```

And for `spin test -s cluster`:
```
ImportError while importing test module '/home/rgommers/code/scipy/build-install/usr/lib/python3.12/site-packages/scipy/cluster/tests/test_hierarchy.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
scipy/cluster/tests/test_hierarchy.py:54: in <module>
    from . import hierarchy_test_data
E   ImportError: cannot import name 'hierarchy_test_data' from 'build-install.usr.lib.python3.12.site-packages.scipy.cluster.tests' (unknown location)
```

Forbidding relative imports in tests is not great. SciPy has quite a few of those.

Reply:
This seems to work smoothly in spin 0.13 now. I think the comment can be resolved, and no more changes should be needed (🤞🏼).